vscode-dotnet-runtime-library/src/Acquisition/VersionResolver.ts

```diff
@@ -1,141 +1,294 @@
+    }
+
+    public async getFullRuntimeVersion(version: string): Promise<string> {
+        return this.getFullVersion(version, true);
+    }
+
+    public async getFullSDKVersion(version: string): Promise<string> {
+        return this.getFullVersion(version, false);
+    }
+
+    /**
+     * @param getRuntimeVersion - True for getting the full runtime version, false for the SDk version.
+     */
+    private async getFullVersion(version: string, getRuntimeVersion: boolean): Promise<string>
+    {
+        const releasesVersions = await this.getReleasesInfo(getRuntimeVersion);
+        return new Promise<string>((resolve, reject) =>
+        {
+            try
+            {
+                const versionResult = this.resolveVersion(version, releasesVersions);
+                this.eventStream.post(new DotnetVersionResolutionCompleted(version, versionResult));
+                resolve(versionResult);
+            }
+            catch (error)
+            {
+                this.eventStream.post(new DotnetVersionResolutionError(error as Error, version));
+                reject(error);
+            }
+        });
+    }
+
+    private resolveVersion(version: string, releases: IDotnetListVersionsResult): string {
+        this.validateVersionInput(version);
+
+        const matchingVersion = releases.filter((availableVersions : IDotnetVersion) => availableVersions.channelVersion === version);
+        if (!matchingVersion || matchingVersion.length < 1)
+        {
+            const err = new DotnetVersionResolutionError(new Error(`The requested and or resolved version is invalid.`), version);
+            this.eventStream.post(err);
+            throw err;
+        }
+
+        return matchingVersion[0].version;
+    }
+
+    private validateVersionInput(version: string) {
+        const parsedVer = semver.coerce(version);
+        if (version.split('.').length !== 2 || !parsedVer)
+        {
+            const err = new DotnetVersionResolutionError(new Error(`An invalid version was requested. Version: ${version}`), version);
+            this.eventStream.post(err);
+            throw err;
+        }
+    }
+
+    private async getReleasesInfo(getRuntimeVersion : boolean): Promise<IDotnetListVersionsResult>
+    {
+        const apiContext: IDotnetListVersionsContext = { listRuntimes: getRuntimeVersion };
+
+        const response = await this.GetAvailableDotnetVersions(apiContext);
+        if (!response)
+        {
+            const err = new DotnetInvalidReleasesJSONError(new Error(`We could not reach the releases API to download dotnet, is your machine offline?`));
+            this.eventStream.post(err);
+            throw err;
+        }
+
+        return response;
+    }
+
+    /**
+     *
+     * @param fullySpecifiedVersion the fully specified version of the sdk, e.g. 7.0.301 to get the major from.
+     * @returns the major.minor in the form of '3', etc.
+     */
+    public static getMajor(fullySpecifiedVersion : string) : string
```

Comment:
Not sure how hard/important it would be to do here, but in general I would steer away from having a lot of static methods. They can be convenient, but can also reduce the number of [seams](https://biratkirat.medium.com/working-effectively-with-legacy-code-changing-software-part-1-chapter-4-b997b78fc0a2), making an already hard to test system even more hard to test :(

Reply:
Thank you for sharing that article on seams, it was a really interesting way of thinking about testing I hadn't thought of :) I went ahead and made them not static